### PR TITLE
Add 'subject_is_issuer' and 'subject_is_holder'

### LINKED ISSUE #49: allow requestor to specify that supplied attributes must be self-attested
as raised in #18 

> TODO: Presumably a request for input that has no constraints is allowed to be self-attested -- but sometimes this isn't good enough. How do I say, "As a university verifying your study habits, I explicitly DON'T want an attestation from your high school guidance counselor; I want you to tell me what you think about yourself?"

### PR DESCRIPTION
**Pull request by brentzundel**
First attempt to draft language that fixes #49 

Signed-off-by: Brent Zundel <brent.zundel@gmail.com>

### CODE REVIEW

File: spec/spec.md [hunk 1]
```diff
@@ -885,6 +885,12 @@ Descriptor Objects_ are composed as follows:
         Setting the property to `false`, or omitting the property, indicates
         the processing entity ****MAY**** submit a response that contains more
         than the data described in the `fields` array.
+      - The object ****MAY**** contain a `self_attested` property, and if
+        present its value ****MUST**** be one of the following strings:
+        - `required` - This indicates that the processing entity ****MUST****
+          submit a response provided by the Holder.
```
Comment (OR13) [inline spec/spec.md:891]: hmm, I'm not sure this language is clear enough... In my mind `self_attested`, means that the issuer of the credential and the holder of the credential are the same... the subject could be anything... I can attest to daniel working for microsoft, and present that credential... in that case.

1. daniel is the subject
2. the claim is microsoft employee
3. the issuer is orie
4. the holder is orie.

compare that to...

1. daniel is the subject
2. the claim is microsoft employee
3. the issuer is pam
4. the holder is orie.

the language needs to be clearer.

**Comment by brentzundel** [comment]
@OR13, added example in [6243122](https://github.com/decentralized-identity/presentation-exchange/pull/84/commits/6243122ba059439ae33b5d9dad3c2b3671250e79). 
Does this clarify things enough, or do you feel the definition itself still need work?

**Comment by OR13** [comment]
> @OR13, added example in [6243122](https://github.com/decentralized-identity/presentation-exchange/pull/84/commits/6243122ba059439ae33b5d9dad3c2b3671250e79).
> Does this clarify things enough, or do you feel the definition itself still need work?

IMO its still not clear.

> Does this introduce a nondeterministic processing step that leaks into the specifics of what every different credential format would require to validate the credential was from a given DID/key?

No, its just about mapping the `self_attested` normative defintion to the vc data model correctly.

Could probably be resolved with text like:

If `self_attested: 'required'` then the issuer of the credential must match the holder making the presentation.... 

This is similar to requirements like "subject===holder" which is "required" for things like "Permanent Resident Cards"... its not valid for me to present your Digital Drivers License... even if the cryptography verifies without exception.

**Comment by wyc** [comment]
I recommend we adopt clearer terminology from our discussion, summarized here: https://github.com/decentralized-identity/presentation-exchange/issues/49#issuecomment-666545366

We can call this property `issuer_is_subject` or `subject_is_issuer`. This eliminates any ambiguity about what is meant by "self-attested", to me you can present a self-attested credential to someone else, such as presenting someone else's self testimony. If we want issuer to be the same as holder, then we should call that one `issuer_is_holder` or `holder_is_issuer`.

**Comment by brentzundel** [comment]
I changed `self_attested` to `subject_is_issuer` and tried to make the explanation of the property more clear.
@OR13 @wyc @csuwildcat let me know what you think.

**Comment by csuwildcat** [comment]
@brentzundel @wyc wait a sec, didn't Wayne illustrate that there were essentially multiple modes under which this will operate? If so, shouldn't the property be more generic, and the value be a string that represents the specific mode of attested state that the Holder must be in relation to the exchange?

**Comment by OR13** [comment]
@csuwildcat overloading a term is generally less safe than being explicit, you would like to add concrete examples to this thread that are not achievable with `subject_is_issuer` ?

**Comment by OR13** [comment]
I'd still prefer to see concrete examples added to the PR, although i do fine the language `subject_is_issuer` much more informative than `self_attested`.

**Comment by csuwildcat** [comment]
> I'd still prefer to see concrete examples added to the PR, although i do fine the language `subject_is_issuer` much more informative than `self_attested`.

Based on the description from Wayne on the call where we talked about this, I was actually thinking we needed something like a prop that talked about the relationship between the holder and the proofs, wherein it could be multiple modes. I may have not understood that correctly though. @wyc can you help clarify for me?

**Comment by wyc** [comment]
So here’s where we left off:
- We have three parties as we’ve seen in the VC spec aside from the verifier: issuer, holder, and subject.
- There are 3 possible relationships among them for the “is” relationship, namely `issuer_is_holder`, `issuer_is_subject`, and `holder_is_subject`, or some semantic equivalent of each.
- The jury is currently out on how to describe relationships across these three non-verifier parties that are more complex than “is” such as “is mother of” or “is authorized by”. This began a discussion about how hubs may or may not be necessary for the verification of these complex relationships to be practical.
- We also with just these flags do not support describing relationships for subjects that are not `credentialSubject.id`, which could be arbitrarily complex and numerous.
- The implementation of the above two checking modes is likely to require expressive complexity that a DSL would be required to satisfy, and we decided this to be out of scope for the editor’s draft.

Hopefully that’s helpful context.

**Comment by brentzundel** [comment]
Rather than having a `subject_is_issuer` property, it may make sense to introduce a meta property which is an array.

The following examples show two ways we could allow a Verifier to request a response from a holder such that the subject must be the holder and the value may be self-attested:

Example 1, using properties:
```
"constraints":{
    "subject_is_issuer": "allowed",
    "subject_is_holder": "required",
    ...
}
```

Example 2, using values for a meta property (I am not committed to any of these property names):
```
"constraints":{
    "role":[
        {
            "role_type": "subject_is_issuer", 
            "role_constraint": "allowed"
        },
        {
            "role_type": "subject_is_holder",
            "role_constraint": "required"
        }
    ],
    ...
}
```

wdyt?

**Comment by csuwildcat** [comment]
@brentzundel prop names aside, I would skew toward the objects for this, given it is something we already know we may want to extend with other props, given a possible need for defining more relationships of greater complexity.

**Comment by brentzundel** [comment]
Consensus of the group during our most recent meeting was for a structure like example 2. 
Pam and I will meet next week to try to align the property names with existing notions from FIDO, then I will update the PR to use those names and the preferred structure.

**Comment by brentzundel** [comment]
PR has been updated according to the latest group consensus. @OR13 @csuwildcat @wyc please review (does anybody know Pam's github handle?)

**Comment by brentzundel** [comment]
> @brentzundel can I ask that you update the json schema and corresponding test to reflect your changes?

I am certainly willing to try, but would need review of my attempts to update it.

**Comment by brentzundel** [comment]
@glcohen attempt has been made in [96a4e4a](https://github.com/decentralized-identity/presentation-exchange/pull/84/commits/96a4e4a0e35c71eba7c3fa0a285b679e1fe457e1) to add `subject_is_issuer` and `subject_is_holder` to the JSON schema and tests, please let me know if it looks right.

**Comment by brentzundel** [comment]
@OR13 about a month ago you requested changes to this PR. Since then many many changes have been made. Could you give feedback as to whether those changes have addressed your concerns?

**Comment by glcohen** [comment]
@OR13 @csuwildcat can we merge?

**Comment by csuwildcat** [comment]
@brentzundel can you recall if any of the recent conversations we had yielded any changes that were desired, or is this ready to go in your opinion?

**Comment by brentzundel** [comment]
all of the changes that were the result of our most recent discussions have been included.

@OR13 requested changes a while back, but hasn't said if the current text addresses his concerns.

**Comment by csuwildcat** [comment]
@brentzundel let's give @OR13 the rest of this PDT morning/early afternoon to comment, then pull it in if there are no further issues.

**Review by OR13 — APPROVED** [review]
Assuming all the normative statements we are making in here will be tested at somepoint, I approve this PR.

**Pull request merged**